File: src/main/java/com/ptmanager/backend/common/ApiExceptionHandler.java
package com.ptmanager.backend.common;

import java.time.Instant;
import java.util.Map;
import java.util.NoSuchElementException;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@RestControllerAdvice
public class ApiExceptionHandler {

    @ExceptionHandler(NoSuchElementException.class)
    public ResponseEntity<ApiError> handleNotFound(NoSuchElementException exception) {
        return ResponseEntity.status(HttpStatus.NOT_FOUND)
            .body(new ApiError("NOT_FOUND", exception.getMessage(), Instant.now(), Map.of()));
    }

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<ApiError> handleBadRequest(IllegalArgumentException exception) {
        return ResponseEntity.badRequest()
            .body(new ApiError("BAD_REQUEST", exception.getMessage(), Instant.now(), Map.of()));
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ApiError> handleValidation(MethodArgumentNotValidException exception) {
        Map<String, String> fields = exception.getBindingResult()
            .getFieldErrors()
            .stream()
            .collect(
                java.util.stream.Collectors.toMap(
                    error -> error.getField(),
                    error -> error.getDefaultMessage() == null ? "invalid" : error.getDefaultMessage(),
                    (left, right) -> left
                )
            );

        return ResponseEntity.badRequest()
            .body(new ApiError("VALIDATION_FAILED", "Request validation failed.", Instant.now(), fields));
    }

    public record ApiError(
        String code,
        String message,
        Instant timestamp,
        Map<String, String> fields
    ) {
    }
}

File: src/main/kotlin/com/ptmanager/backend/common/ApiExceptionHandler.kt
package com.ptmanager.backend.common

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.multipart.MaxUploadSizeExceededException
import org.springframework.web.server.ResponseStatusException
import java.time.Instant
import java.util.NoSuchElementException

@RestControllerAdvice
class ApiExceptionHandler {

    @ExceptionHandler(NoSuchElementException::class)
    fun handleNotFound(exception: NoSuchElementException): ResponseEntity<ApiError> =
        ResponseEntity.status(HttpStatus.NOT_FOUND)
            .body(ApiError("NOT_FOUND", exception.message, Instant.now(), emptyMap()))

    @ExceptionHandler(IllegalArgumentException::class)
    fun handleBadRequest(exception: IllegalArgumentException): ResponseEntity<ApiError> =
        ResponseEntity.badRequest()
            .body(ApiError("BAD_REQUEST", exception.message, Instant.now(), emptyMap()))

    @ExceptionHandler(ResponseStatusException::class)
    fun handleResponseStatus(exception: ResponseStatusException): ResponseEntity<ApiError> {
        val status = exception.statusCode
        val code = (status as? HttpStatus)?.name ?: status.toString()
        return ResponseEntity.status(status)
            .body(ApiError(code, exception.reason, Instant.now(), emptyMap()))
    }

    @ExceptionHandler(MaxUploadSizeExceededException::class)
    fun handlePayloadTooLarge(exception: MaxUploadSizeExceededException): ResponseEntity<ApiError> =
        ResponseEntity.status(HttpStatus.PAYLOAD_TOO_LARGE)
            .body(ApiError("PAYLOAD_TOO_LARGE", "파일 용량이 초과되었습니다.", Instant.now(), emptyMap()))

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidation(exception: MethodArgumentNotValidException): ResponseEntity<ApiError> {
        val fields = exception.bindingResult.fieldErrors.associate { error ->
            error.field to (error.defaultMessage ?: "invalid")
        }
        return ResponseEntity.badRequest()
            .body(ApiError("VALIDATION_FAILED", "Request validation failed.", Instant.now(), fields))
    }

    data class ApiError(
        val code: String,
        val message: String?,
        val timestamp: Instant,
        val fields: Map<String, String>,
    )
}
